### ISSUE TITLE
WebView2 memory leak

### ISSUE THREAD

**Issue by tj-jyk** [MicrosoftEdge/WebView2Feedback#1633]
**Description**

WebView2 (and old WebView) exhibits a memory leak when creating and deleting the control in a loop. In my application, the user can change the content every (x) seconds and with a WebView/WebView2 control the user will get a memory leak. I opened already 2 Issues in [WindowsAppSDK repo](https://github.com/microsoft/WindowsAppSDK/issues/1158) and in [WinUI repo](https://github.com/microsoft/microsoft-ui-xaml/issues/5622)

**Version**
SDK: 1.0.902.49
Runtime: 92.0.902.62 
Framework: UWP
OS: Win10  10.0.19042 Build 19042

**Repro Steps**
[UWP Project GitHub repo](https://github.com/tj-jyk/WebView_WebView_memory_leak)

1. Start a loop / a timer `var timer = new DispatcherTimer
			{
				Interval = TimeSpan.FromSeconds(1)
			};
			timer.Tick += Timer_Tick;
			timer.Start();`
2.  In a loop logic find the old WebView/WebView2 control in a DOM `VisualTreeHelper.GetChildrenCount(parent);  ..... VisualTreeHelper.GetChild(parent, i) ...`)
3.  Remove the old WebView from the DOM `(VisualTreeHelper.DisconnectChildrenRecursive(rootGrid);)`
4.  For a WebView call `webViewToRemove.Navigate(new Uri("about:blank"));` For a WebView2:  `webViewToRemove.Close(); `
5. Add a new WebView/WebView2 control to the DOM `rootGrid.Children.Add(newWebView);`
6. Application crashes after 15 minutes.

**Screenshots**
![image](https://user-images.githubusercontent.com/29145276/127843163-d17ed5a5-18a8-4c60-bd2d-e25d30f7a497.png)

**Additional context**

[AB#35385513](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/35385513)

**Comment by champnic** [comment]
Thanks for the bug report @tj-jyk! I've added it to our backlog to take a look.

In the meantime do you need to remove and create a new WebView2 every time? Can you reuse the WebView2 when you need to change content?

**Comment by tj-jyk** [comment]
Hi @champnic. 
Yes, it's possible and its work. But if the component has a memory leak, will be better to get an update, otherwise you will have to spend a lot of time rewriting the existing code. I found many articles in the internet and nobody has a right answer.

**Comment by champnic** [comment]
I completely agree we should try and fix this memory leak - I just wanted to make sure that your app wasn't blocked by this issue :)

**Comment by AdamCaviness** [comment]
In my case it appears a leak stems from the HwndSourceKeyboardInputSite. I am calling Dispose() but the WebView2 instance remains in memory.

![image](https://user-images.githubusercontent.com/534790/142268887-df7b9e49-9dd5-4cba-aaae-42a6120e9847.png)

**Comment by champnic** [comment]
Hey @AdamCaviness - Can you share how you are hooking up to the WebView2 and how/when you are calling Dispose?

**Comment by AdamCaviness** [comment]
@champnic Gladly. In my scenario, the WebView2 component is not placed in a dedicated window  but is instead shown directly in the MainWindow with an active screen approach. The sample ([WpfWebView2Leak](https://github.com/AdamCaviness/WpfWebView2Leak)) demonstrates the same single window app approach. The WebView2 component is disposed in the WebScreen Close button handler. The derived WebView2 approach is intended to address a Windows sleep issue with WebView2. I used dotMemory to profile the sample. Thanks.

**Comment by champnic** [comment]
Thanks for sharing. I've verified that the WebView2 processes and most of the WebView2 code is being cleaned up at least, as those are where most of the memory usage would be. I wonder if there's a bug in WPF around making sure the HwndSourceKeyboardInputSite gets removed from the list. We'll take a closer look. Thanks!

**Comment by MatevzRopret** [comment]
I'd just like to report that we have found the same issue: attaching the Webview to the UI is what causes the leak.
If the webview is not attached to the UI it won't leak.

Here is an example project (VS solution)

[Webview2MemoryLeakTest.zip](https://github.com/MicrosoftEdge/WebView2Feedback/files/7853836/Webview2MemoryLeakTest.zip)

**Comment by fabioparodi** [comment]
I have the same problem. 
I've modified a very old and stable C++/MFC application to use WebVIew2. Then scripted to create-show-hide-dispose the WebVIew2 window in a cycle every 10 seconds. I can see the proces private bytes increase, it is very regular.
I am going to pool and reuse the WebView2 windows - never ever dispose.
Still an open bug

**Comment by fabioparodi** [comment]
![image](https://user-images.githubusercontent.com/8512210/178445920-584105aa-ccd7-4b7e-96f5-87a504f55b47.png)

**Comment by fabioparodi** [comment]
![image](https://user-images.githubusercontent.com/8512210/178446134-b5fd4732-dc5a-43fb-a411-81aa9f832824.png)

**Comment by juipinjao** [comment]
fanioparoid
I face the same problem on a MFC app. too. Just wonder if you find any solution? Thanks

**Comment by agenne** [comment]
The original issue posted by @tj-jyk (my colleague) seems to be solved. 
Updated the WebView2 version to WinUI 2.8.1.
It is still necessary to cleanup, but it is sufficient to call Close on the webview which is much better than navigate to about:blank and call Gc.Collect

**Comment by champnic** [comment]
Glad to hear it's working now @agenne!

For others on this thread - if you are still seeing the issue and it's not in WinUI 2.8.1 or WindowsAppSDK, please open a separate issue for us to track. Thanks!
